### PR TITLE
Added FastCheck Filters #minor

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
 - Fast check filters allow to look for existence of an item and prioritize false negatives instead of false positives like Bloom filters do.
False negative here refers to the filter returning a negative, even if it is possible that the object was previously seen, instead of false for an object that was never seen, as in the case of bloom filters.

- The code had to be adapted from https://github.com/jmhodges/opposite_of_a_bloom_filter/blob/master/go/oppobloom/oppobloom.go because, the existing implementation had only one method that would check and add. We prefer the case in which can check and then add as 2 separate invocations

Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue

https://github.com/flyteorg/flyte/issues/1615
https://github.com/flyteorg/flyte/issues/1614

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by hamersaw** [comment]
The benchmarking is currently comparing oppoBloom vs oppoBloom, we never initialize a lruCache instance. I added some additional change to break it up a bit (split Contains / Add benchmarks, use pb.Next for a single operation, etc) below:

```
func benchmarkFilter(b *testing.B, ctx context.Context, ids [][]byte, f func(context.Context, []byte) bool, parallelism int) {
	b.SetParallelism(parallelism)
	b.ResetTimer()

	b.RunParallel(func(pb *testing.PB) {
		index := 0
		for pb.Next() {
			index = (index + 1) % len(ids)
			f(ctx, ids[index])
		}
	})
}

func BenchmarkFilter(b *testing.B) {
	parallelism := 10
	cacheSize := 5000

	ctx := context.TODO()
	bf, err := NewOppoBloomFilter(cacheSize, promutils.NewTestScope())
	assert.NoError(b, err)
	lf, err := NewLRUCacheFilter(cacheSize, promutils.NewTestScope())
	assert.NoError(b, err)
	ids := RandBytesSliceGenerator(100, 1000000)

	b.Run("oppoFilterContainsBenchmark", func(b *testing.B) {
		benchmarkFilter(b, ctx, ids, bf.Contains, parallelism)
	})

	b.Run("oppoFilterAddBenchmark", func(b *testing.B) {
		benchmarkFilter(b, ctx, ids, bf.Add, parallelism)
	})

	b.Run("lruCacheContainsBenchmark", func(b *testing.B) {
		benchmarkFilter(b, ctx, ids, lf.Contains, parallelism)
	})

	b.Run("lruCacheAddBenchmark", func(b *testing.B) {
		benchmarkFilter(b, ctx, ids, lf.Add, parallelism)
	})
}
```

Playing with parallelism, cache size, and id size shows some more interesting things.

```
# parallelism 1
BenchmarkFilter/oppoFilterContainsBenchmark-12           9239407           130 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               18759355            60.4 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12            10047250           125 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1773 ns/op

# parallelism 50
BenchmarkFilter/oppoFilterContainsBenchmark-12           8767063           143 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               20322484            64.9 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12             8519787           134 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1227 ns/op

# parallelism 100
BenchmarkFilter/oppoFilterContainsBenchmark-12           9021384           139 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               21777622            64.0 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12             8483595           135 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1144 ns/op



# id size 10 (parallelism 10, cache size 5000)
BenchmarkFilter/oppoFilterContainsBenchmark-12           7281070           170 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               19354623            58.7 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12             8434303           148 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1266 ns/op

# id size 50 (parallelism 10, cache size 5000)
BenchmarkFilter/oppoFilterContainsBenchmark-12           8172189           161 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               22396275            54.0 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12             9730503           134 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1278 ns/op

# id size 100 (parallelism 10, cache size 5000)
BenchmarkFilter/oppoFilterContainsBenchmark-12           8134311           147 ns/op
BenchmarkFilter/oppoFilterAddBenchmark-12               18416812            64.6 ns/op
BenchmarkFilter/lruCacheContainsBenchmark-12             9610022           132 ns/op
BenchmarkFilter/lruCacheAddBenchmark-12                  1000000          1313 ns/op
```

For instance, contains queries are very similar between the two (unsurprising as the both just use a hash lookup), but regardless of the parameters Add operations are 2 orders of magnitude faster than oppoFilter (which is more aligned with what we expected).
